Advanced-Recruit-Training/src/BatteryStateOfChargeService.cpp

```diff
@@ -16,23 +27,51 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+
+    return totalAmpHour_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
+
+    if(current_<0){
+       return true;
+    }
     return false;
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
-{
-    return QTime::currentTime();
+{    QTime charging;
+     charging.setHMS(0,0,0,0);
+
+    if(isCharging()){
+    charging= charging.addMSecs(qAbs(totalAmpHour_/current_*MS_TO_HOURS));
+
+     return charging;
+    }
+
+   charging = charging.addMSecs(qAbs(( BATTERY_AMP_HOUR_CAPACITY-totalAmpHour_)/current_*MS_TO_HOURS));
+    return charging;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
 {
     Q_UNUSED(batteryData);
-    // This is where you can update your variables
-    // Hint: There are many different ways that the totalAmpHoursUsed can be updated
-    // i.e: Taking a running average of your data values, using most recent data points, etc.
+
+    voltage_ = batteryData.voltage;
+    current_ = batteryData.current;
+
+    hour_= batteryData.time.hour()
+            +batteryData.time.minute()/HOUR_TO_M
+            +batteryData.time.second()/S_TO_HOURS
+            +batteryData.time.msec()/MS_TO_HOURS;
+    //using running average of current to calculate amp hours
+    if(counter_==0){    //the first hour read is stored in prevhour
```

Comment:
Since this is only true once, I would recommend using a boolean variable instead.